xen/arch/arm/coproc/coproc.c

```diff
@@ -73,162 +69,293 @@ static struct coproc_device *coproc_find_by_path(const char *path)
+                                                    struct mcoproc_device *mcoproc,
+                                                    const struct dt_device_node *np_vcoproc)
 {
     struct vcoproc_instance *vcoproc;
-    int ret;
+    int ret = 0;
+    int i;
 
     vcoproc = xzalloc(struct vcoproc_instance);
     if ( !vcoproc )
     {
         printk("Failed to allocate vcoproc_instance for %s\n",
-               dev_path(coproc->dev));
+               dev_path(mcoproc->dev));
         return ERR_PTR(-ENOMEM);
     }
 
-    vcoproc->coproc = coproc;
+    vcoproc->mcoproc = mcoproc;
     vcoproc->domain = d;
     vcoproc->state = VCOPROC_UNKNOWN;
     spin_lock_init(&vcoproc->lock);
 
-    ret = coproc->ops->vcoproc_init(vcoproc);
-    if ( ret )
+    vcoproc->num_mmios = dt_number_of_address(np_vcoproc);
+
+    if ( vcoproc->num_mmios != mcoproc->num_mmios )
     {
-        printk("Failed to initialize vcoproc_instance for %s\n",
-               dev_path(coproc->dev));
+        printk("MMIO num mistmatch for \"%s\", %u vs %u\n",
+               dt_node_full_name(np_vcoproc), vcoproc->num_mmios, mcoproc->num_mmios);
+        ret = -ENODEV;
         goto out;
     }
 
-    spin_lock(&coproc->vcoprocs_lock);
-    list_add(&vcoproc->vcoproc_elem, &coproc->vcoprocs);
-    spin_unlock(&coproc->vcoprocs_lock);
-
-    return vcoproc;
-
-out:
-    xfree(vcoproc);
-    return ERR_PTR(ret);
-}
+    vcoproc->mmios = xzalloc_array(struct vcoproc_mmio, vcoproc->num_mmios );
+    if ( !vcoproc->mmios )
+    {
+        printk("Failed to allocate %d mmio(s) for \"%s\"\n", vcoproc->num_mmios,
+               dev_path(mcoproc->dev));
+        ret = -ENOMEM;
+        goto out;
+    }
 
-struct vcoproc_instance *coproc_get_vcoproc(struct domain *d,
-                                            struct coproc_device *coproc)
-{
-    struct vcoproc_instance *vcoproc = NULL;
-    bool_t found = false;
-    unsigned long flags;
+    for ( i =0; vcoproc->num_mmios > i; i++)
+    {
+        struct mcoproc_mmio *m_mmio = &mcoproc->mmios[i];
+        struct pcoproc_mmio *p_mmio = m_mmio->p_mmio;
+        u64 size;
+        int index = dt_property_match_string(np_vcoproc, "reg-names", p_mmio->name);
 
-    spin_lock_irqsave(&coproc->vcoprocs_lock, flags);
+        vcoproc->mmios[i].vcoproc = vcoproc;
 
-    if ( list_empty(&coproc->vcoprocs) )
-        goto out;
+        if (index < 0)
+        {
+            if ( vcoproc->num_mmios > 1 )
+            {
+                printk("More than one mmio and failed name match for %s\n",
+                       dev_path(mcoproc->dev));
+                goto out;
+            }
+            else
+                index = 0;
+        }
 
-    list_for_each_entry( vcoproc, &coproc->vcoprocs, vcoproc_elem )
-    {
-        if ( vcoproc->domain == d )
+        ret = dt_device_get_address(np_vcoproc, index,
+                                    &vcoproc->mmios[i].addr,
+                                    &size);
+        if ( ret )
```

Comment:
could you please point me to extra spaces? If you are talking about spaces around bracers in the if condition - they are needed by xen coding style, see https://github.com/xen-troops/xen/blob/xt-master-wip/CODING_STYLE#L43. The question is that I missed them few lines below